bitcoin/src/address.rs

```diff
@@ -554,8 +564,108 @@ impl<'a> fmt::Display for AddressEncoding<'a> {
     }
 }
 
+mod sealed {
+    pub trait NetworkValidation {}
+    impl NetworkValidation for super::NetworkChecked {}
+    impl NetworkValidation for super::NetworkUnchecked {}
+}
+
+/// Marker of status of address's network validation. See section [*Parsing addresses*](Address#parsing-addresses)
+/// on [`Address`] for details.
+pub trait NetworkValidation: sealed::NetworkValidation {}
+
+/// Marker that address's network has been successfully validated. See section [*Parsing addresses*](Address#parsing-addresses)
+/// on [`Address`] for details.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum NetworkChecked {}
+
+/// Marker that address's network has not yet been validated. See section [*Parsing addresses*](Address#parsing-addresses)
+/// on [`Address`] for details.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum NetworkUnchecked {}
+
+impl NetworkValidation for NetworkChecked {}
+impl NetworkValidation for NetworkUnchecked {}
+
 /// A Bitcoin address.
 ///
+/// ### Parsing addresses
```

Comment:
While reviewing I wrote some code, you can use any of this in the examples if you like it
```rust
    // Use `FromStr` to get an `Address` from a string.
    let address = Address::from_str("32iVBEu4dxkUQk9dJbZUiBiQdmypcEyJRf")
        .expect("failed to parse address")
        .require_network(Network::Bitcoin)
        .expect("wrong network");

    // Display or serialize an `Address`.
    println!("address: {}", address);
    let ser = serde_json::to_string(&address).expect("failed to serialize address");

    // Use `parse` to get an unchecked address from a string.
    let address: Address<NetworkUnchecked> = "32iVBEu4dxkUQk9dJbZUiBiQdmypcEyJRf"
        .parse()
        .expect("failed to parse address");

    // `Display` is explicitly *not* implemented for `Address<NetworkUnchecked>`.
    println!("address: {:?}", address);
    
    // `Serialize` is also explicitly *not* implemented for `Address<NetworkUnchecked>`.
    // let _ = serde_json::to_string(&address).expect("failed to serialize address");

    // The generic on `Address<T>` is there to catch incorrect addresses and to prevent hard coding of the network in bitcoin codebases - ouch current API does not encourage _not_ hard coding the network.
    let address = address
        .require_network(Network::Bitcoin)
        .expect("wrong network");

    // Equivalent to
    let address: Address<NetworkChecked> = "32iVBEu4dxkUQk9dJbZUiBiQdmypcEyJRf"
        .parse::<Address<NetworkUnchecked>>()
        .expect("failed to parse address")
        .require_network(Network::Bitcoin)
        .expect("wrong network");
```